Implement more copying APIs in pylibcudf

vyasr opened the pull request:
## Description
<!-- Provide a standalone description of changes in this PR. -->
<!-- Reference any issues closed by this PR with "closes #1234". -->
<!-- Note: The pull request title will be included in the CHANGELOG. -->
This PR implements a handful of libcudf copying APIs in pylibcudf and rewrites cudf Cython internals to use those functions where appropriate. This PR only implements those methods that can be done without any additional changes to the pylibcudf Column class. I'll follow up with other PRs for changes that require those changes.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

vuule commented on cpp/include/cudf/copying.hpp:
```diff
@@ -211,7 +211,7 @@ std::unique_ptr<table> scatter(
  * @brief Indicates when to allocate a mask, based on an existing mask.
  */
-enum class mask_allocation_policy {
+enum class mask_allocation_policy : int32_t {
```
why not a smaller type?

vyasr replied:
No particular reason other than that's what our other enums use. I think everything has probably just copied `type_id`: https://github.com/rapidsai/cudf/blob/branch-24.02/cpp/include/cudf/types.hpp#L187

vuule reviewed (approved):
C++ looks good :D

shwina commented on python/cudf/cudf/_lib/pylibcudf/copying.pyx:
```diff
@@ -55,3 +66,198 @@ cpdef Table gather(
+    return Table.from_libcudf(move(c_result))
+
+
+cpdef object empty_like(object input):
```
Is it worth having an `empty_column_like()` and `empty_table_like()`?

vyasr replied:
I debated this as well. I could go either way as long as whatever we decide here is made consistent policy across pylibcudf. The current approach matches the libcudf API and makes it simpler for the user at the cost of having to deal with typing internally and introducing a bit of overhead as a result. The alternative would be strongly typed but slightly more verbose functions. Do you think the latter is better?

shwina replied:
I think:

- Separate APIs for `Column`/`Table`
- No need for separate APIs for other kinds of overloads in C++

vyasr replied:
I'll also include scalars alongside column and tables, but otherwise yes agreed.

vyasr replied:
Hmm this choice ended up being a bit more loaded than I would have liked given the functions that accept either a table or a vector of scalars. Let me know what you think of the current set of functions that I've defined.

shwina replied:
Hmm, I agree I don't love it. If this were pure Python maybe `overload`s would be the right choice typing wise.

vyasr replied:
We discussed naming offline. The state of the PR as of https://github.com/rapidsai/cudf/pull/14508/commits/b0ae7f0ff7ca31c35b84e0e87723450f16bc1b69 is our final choice of names.

wence- commented on python/cudf/cudf/_lib/pylibcudf/copying.pyx:
```diff
@@ -1,23 +1,34 @@
 from .table cimport Table
 
+# workaround for https://github.com/cython/cython/issues/3885
+ctypedef const scalar constscalar
```
That bug is now fixed, so plausibly we don't need this any more. But I think you need this for cython/cython#4180, so **suggestion** update the issue link?

vyasr replied:
Ah good catch, yes I documented that in https://github.com/rapidsai/cudf/issues/14023 but didn't update this link. Can do.

wence- commented on python/cudf/cudf/_lib/pylibcudf/copying.pyx:
```diff
@@ -55,3 +66,198 @@ cpdef Table gather(
+                )
+            )
+    elif isinstance(source, list):  # TODO: is list too restrictive?
+        for slr in source:
+            source_scalars.push_back(
+                reference_wrapper[constscalar](dereference(slr.c_obj))
+            )
```
**suggestion** This pattern already repeats twice, let's pull it into a utility function.

vyasr replied:
Done.

wence- commented on python/cudf/cudf/_lib/pylibcudf/copying.pxd:
```diff
@@ -13,3 +15,23 @@ cpdef Table gather(
+    size_type input_begin,
+    size_type input_end,
+    size_type target_begin,
+)
```
**Suggestion/nitpick** What order should write these? I would suggest one of:

1. Alphabetical
2. The same order as the C++ header file
3. The same order as the doxygen documentation of the matching header file (which I think is the same as (2)

Ideally we would also name the arguments the same, but that is very persnickity.

vyasr replied:
I meant to do that, thanks for the reminder. I think matching the header makes sense.

vyasr replied:
Done.

vyasr commented on python/cudf/cudf/_lib/copying.pyx:
```diff
@@ -598,45 +467,33 @@ def boolean_mask_scatter(list input_, list target_columns,
-            target_columns,
-            boolean_mask
-        )
+        plc_input = pylibcudf.Table([col.to_pylibcudf(mode="read") for col in input_])
```
It's not reserved per se. It's a built-in function as opposed to a keyword like `in`, so you can shadow it. It's only really an issue if you actually needed that function, but I don't know if it might confuse some linters (I haven't observed that in a long time though) and it can lead to odd syntax highlighting in editors. I can change it if you'd like, no strong preference.

vyasr replied:
Also no longer relevant as above.

vyasr commented on python/cudf/cudf/_lib/cpp/copying.pxd:
```diff
@@ -36,118 +36,118 @@ cdef extern from "cudf/copying.hpp" namespace "cudf" nogil:
-        table_view source_table,
-        column_view scatter_map,
-        table_view target_table,
+        const table_view source_table,
```
Correct, unfortunately not supported.

vyasr commented on python/cudf/cudf/_lib/cpp/copying.pxd:
```diff
@@ -36,118 +36,118 @@ cdef extern from "cudf/copying.hpp" namespace "cudf" nogil:
-        column_view lhs,
-        scalar rhs,
-        column_view boolean_mask
+        const column_view lhs,
```
Good call, I'll fix that.

vyasr commented:
I've addressed all the comments but still need to track down the dask test failure. It doesn't actually look dask-related, but rather a case that's better covered by our dask tests than our other tests.

vyasr commented:
Whoops sorry @vuule I clicked on the wrong button and re-requested you unintentionally!

wence- reviewed (approved):
Minor (nonblocking) suggestions, LGTM!

vyasr commented on python/cudf/cudf/_lib/pylibcudf/copying.pyx:
```diff
@@ -1,23 +1,51 @@
+
+    cdef vector[reference_wrapper[const scalar]] c_scalars
+    c_scalars.reserve(len(source))
+    cdef Scalar slr
+    for slr in source:
+        c_scalars.push_back(
+            reference_wrapper[constscalar](dereference(slr.c_obj))
+        )
```
I did not know that syntax existed! Cool, yes that works better.

vyasr commented on python/cudf/cudf/_lib/pylibcudf/copying.pyx:
```diff
@@ -55,3 +83,187 @@ cpdef Table gather(
+
+
+cpdef Column allocate_like(
+    Column input_column, mask_allocation_policy policy, size=None
```
Hmm this raises a good question of in what ways we want to deviate from libcudf APIs. Keyword-only arguments could be nice, but should we have systematic rules on when we use them?

wence- replied:
My preference is that optional arguments should be keyword-only, which is perhaps an easy systematic rule.

My rationale for this is that multiple defaulted keyword-like (but also positional) arguments are easy to get wrong when using an API. For example, `allocate_like` here takes an overriding `size`, but what if it also took an overriding `dtype`. With `allocate_like(input, policy, size=None, dtype=None)` I have to remember which way round to specify things if I want to override both, with `allocate_like(input, policy, *, size=None, dtype=None)` I don't, and can't get it wrong.

Equally, the use of the keyword name makes the (often untyped) argument self-documenting. It's obvious at the call-site that we're passing a mask_allocation_policy in as the policy argument (since that's an enum that we have to get a value out of), but for `size` it could come from anywhere.

vyasr replied:
I agree, and @shwina also agreed offline. Unfortunately, Cython doesn't support keyword-only arguments for cdef functions, and therefore also not for cpdef functions.

vyasr commented on python/cudf/cudf/_lib/types.pyx:
```diff
@@ -270,9 +271,13 @@ cpdef dtype_to_pylibcudf_type(dtype):
-        SUPPORTED_NUMPY_TO_PYLIBCUDF_TYPES[np.dtype(dtype)]
-    )
+
+    # libcudf types don't support localization so convert to the base type
```
It is, but I don't think it matters. At least, it's equivalent to what we were doing previously. The reason `dtype_to_data_type` doesn't have this is because we do implicit preprocessing by calling [`_get_base_dtype`](https://github.com/rapidsai/cudf/blob/branch-24.02/python/cudf/cudf/_lib/column.pyx#L338), which [does the same thing](https://github.com/rapidsai/cudf/blob/branch-24.02/python/cudf/cudf/utils/dtypes.py#L748).

vyasr commented:
Approved offline with the current set of changes, going to merge now.

vyasr commented:
/merge
